docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,670 @@
+RowAccess(m) ==
+    m.a > 0         \* should flag a type error
+```
+
+**Blocker.** It is not clear to me, what type we would assign to the record
+access operator. Although the type of `m` in `FieldAccess` is obvious to a
+human reader, as it simply requires us to do the top-to-the-bottom type
+propagation, Snowcat constructs a set of equality constraints that are solved
+by unification. This approach requires that we capture the record access
+operator `.` as an equality over type variables and types.
+
+### 4.3. Support plain records and variants, including row typing
+
+This solution follows the approach outlined by [Leijen05][], except that we
+require field names to be unique, not scoped. We present our approach in the
+next section.
+
+## 5. Solution
+
+In the following, we present row types as type terms. We discuss the
+user-facing syntax of the type system later in the text.
+
+### 5.1. Plain records
+
+By using [Row types][], we should be able to infer the type of `m` in
+`RowAccess`:
+
+```
+Rec(RowCons("a", Int, z))
+```
+
+In this example, `RowCons("a", Int, z)` is a type term indicating that the type
+has the field called `a`, and its type is `Int`. On top of that, this row
+extends a parametric type `z`, which either contains a non-empty sequence of
+rows, or is an empty sequence, that is, `RowNil`. Importantly, `RecCons` is
+wrapped with the term `Rec`, so no additional fields can be added to the type.
+
+The example `FieldAccess` contains a record constructor
+`[ a |-> 2, b |-> "B" ]`. We can write a general type inference rule for it:
+
+```
+e_1: t_1, ..., e_n: t_n
+-------------------------------- [rec]
+[ f_1 |-> e_1, ..., f_n |-> e_n]:
+  Rec(
+    RowCons(f_1, t_1,
+      RowCons(f_2, t_2,
+        ...
+          RowNil)
+      )...)
+  )
+```
+
+In `FieldAccess`, we use row types to construct a series of type equations
+(over free type variables `k`, ..., `q`):
+
+```
+// from LET m == [ a |-> 2, b |-> "B" ] IN
+m_type = Rec("a", Int, RowCons("b", Str, RowNil))
+// from m.a > 1
+m_type = Rec(RowCons("a", k, l))
+k = Int
+// from m.b = "B"
+m_type = Rec(RowCons("b", m, n))
+m = Str
+// from m.c = { 1, 2 }
+m_type = Rec(RowCons("c", p, q))
+p = Set(Int)
+```
+
+To solve the above equations, one has to apply unification rules. Precise
+unification rules for rows are given in the paper by [Leijen05][]. Importantly,
+their unification rules allow `RowCons(f1, t1, RowCons(f2, t2, r3))` to be
+unified with `RowCons(f2, t2, RowCons(f1, t1, r3))`. Hence, fields may bubble
+up to the head, and it should be possible to isolate a single field and assign
+the rest to a type variable. By partially solving the above equation, we would
+arrive at contradiction. This supports our intuition that the operator
+`FieldAccess` is ill-typed.
+
+Hence, we formulate the type inference rule for record access in our type
+system as follows:
+
+```
+r: Rec(RowCons("f", t_1), t_2)
+-------------------------------- [rec_acc]
+r.f: t_1
+```
```

Comment:
You'd also need a rule
```
r: RowCons("f", t_1, t_2)
-------------------------------- [xyz]
r: t2
```

Reply:
Why? You would not. The trick is that unification is allowed to swap pairs in rows.

Reply:
>The trick is that unification is allowed to swap pairs in rows.

Where is this mentioned?